# Take a Peek into Belgium's Real Estate Market

The aim of the writing of this Jupyter Notebook is to provide insight on Belgium's real estate through the analysis of data through the scraping of [ImmoWeb](https://www.immoweb.be/), Belgium's prémiere real estate website with the self-acclaimed largest repository of classified real estate listings.

In this occasion, the author focuses on listings that are classified as "for-sale" listing.

#### Step 1 : Data Cleaning

To start data cleaning, we should open the dataset first and convert it into a pandas.DataFrame object in order to allow ourselves the possibility to examine the data in the dataset further. 

The dataset that I am going to use in this occasion is obtained through the execution of the script produced by the end of "Collecting Data Challenge", which is on June 30th 2023. The dataset is called "final-csv.csv", which is a Comma-Separated-Value (CSV) file that you can find in the same directory of this Jupyter Notebook.

In [12]:
####################################################################################### INITIALIZING CURRENT_DIR ################################################################################
import os

def CURRENT_DIR():
    cwd = os.getcwd()
    return os.chdir(cwd[:(cwd.index("Eliza")+5)])

CURRENT_DIR()

print(os.getcwd())

c:\Users\dimas\Desktop\BECODE\PROJECTS\Project_MLDeployment_ImmoEliza


In [13]:
# IMPORTING LIBRARIES
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [14]:
# OPENING [final-csv.csv]
with open("data/final-csv.csv", "r+") as source_csv:
    df_main = pd.read_csv(source_csv)
    pass

df_main.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15283 entries, 0 to 15282
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   URL                    15283 non-null  object 
 1   Listing_ID             15283 non-null  int64  
 2   Type                   15283 non-null  object 
 3   Subtype                15283 non-null  object 
 4   Price                  15283 non-null  object 
 5   Bedroom                14895 non-null  float64
 6   Living_area            13662 non-null  float64
 7   Listing_address        15283 non-null  object 
 8   Postal_code            15283 non-null  object 
 9   Locality               15283 non-null  object 
 10  District               15101 non-null  object 
 11  Swimming_pool          15283 non-null  bool   
 12  Garden                 5088 non-null   object 
 13  Garden_area            2894 non-null   float64
 14  Surface_of_land        9229 non-null   float64
 15  Te

From the DataFrame above, it can be inferred that df_main, (which is loaded from final-csv.csv) has 15283 entries structured through 21 columns. However, there are some datas we have to get rid of in order to allow the possibility of a factual analysis to take place. 

First, let's deal with duplicated entries

The DataFrame below shows every unique instance of duplicated entries in "df_main", which is amounted to 6141 duplicates. It uses the entries in "df_main['Listing_ID']" as unique identifier to filter duplicated entries out, since each classified listing is assigned to a unique 'Listing_ID'

In [15]:
df_main[df_main.duplicated(subset=["Listing_ID"]) == True].sort_values(by="Listing_address")

URL  Listing_ID  \
12397  https://www.immoweb.be/en/classified/villa/for...    10658145   
10273  https://www.immoweb.be/en/classified/house/for...    10554963   
15194  https://www.immoweb.be/en/classified/house/for...    10649511   
11932  https://www.immoweb.be/en/classified/apartment...    10659742   
12241  https://www.immoweb.be/en/classified/house/for...    10658639   
...                                                  ...         ...   
11005  https://www.immoweb.be/en/classified/house/for...    10663324   
14164  https://www.immoweb.be/en/classified/apartment...    10653157   
9623   https://www.immoweb.be/en/classified/house/for...    10668305   
15148  https://www.immoweb.be/en/classified/house/for...    10649653   
13512  https://www.immoweb.be/en/classified/villa/for...    10654779   

            Type          Subtype    Price  Bedroom  Living_area  \
12397      house            villa  1300000      3.0        353.0   
10273      house            house   410000      3.0        150.0   
15194      house            house   620000      4.0        141.0   
11932      house  apartment block   285000      6.0        250.0   
12241      house            house   139000      2.0          NaN   
...          ...              ...      ...      ...          ...   
11005      house            house   220000      4.0        125.0   
14164  apartment        apartment   250000      2.0          NaN   
9623       house            house   685000      5.0        340.0   
15148      house            house   199000      4.0        291.0   
13512      house            villa  3500000      6.0        277.0   

                Listing_address Postal_code          Locality  ...  \
12397             'T Helleke 12        8500          Kortrijk  ...   
10273        's Hertogendijk 95        2940          Stabroek  ...   
15194  's-Hertogenwijngaard 49,        3000            Leuven  ...   
11932                   ... ...        5500  Dinant Anseremme  ...   
12241                       / /        7350           Hensies  ...   
...                         ...         ...               ...  ...   
11005         rur Emile Albot 3        7900         Grandmetz  ...   
14164     sint lambertuslaan 42        3740            Bilzen  ...   
9623         stationsstraat 101        9690       Kluisbergen  ...   
15148  stationsstraat 8, gistel        8470            Gistel  ...   
13512              villapad 12,        8300            Knokke  ...   

      Swimming_pool  Garden Garden_area  Surface_of_land  Terrace  \
12397         False     NaN         NaN           1661.0     True   
10273         False    True         NaN            447.0     True   
15194         False    True        82.0            241.0     True   
11932         False     NaN         NaN              1.0     True   
12241         False     NaN         NaN            138.0     True   
...             ...     ...         ...              ...      ...   
11005         False    True         NaN            360.0     True   
14164         False     NaN         NaN              NaN     True   
9623          False     NaN         NaN           2444.0     True   
15148         False     NaN         NaN            268.0     True   
13512         False     NaN         NaN            588.0     True   

             Kitchen Facade  Open Fire  Furnished State of the building  
12397      installed    4.0      False      False                AS_NEW  
10273      installed    3.0       True        NaN                  GOOD  
15194      installed    2.0      False      False                  GOOD  
11932  semi equipped    2.0      False      False           TO_RENOVATE  
12241      installed    2.0      False      False                  GOOD  
...              ...    ...        ...        ...                   ...  
11005      installed    2.0      False      False                  GOOD  
14164      installed    NaN      False        NaN                  GOOD  
9623   semi equipped    NaN       T

The removal of the duplicates listed above can be done. The script below will initiate the removal process.

In [16]:
df_main_filtered = df_main.sort_values(
    by=["Listing_address"], ascending=True).drop_duplicates(
        subset=["Listing_ID"], keep="last")
df_main_filtered


URL  Listing_ID  \
4930   https://www.immoweb.be/en/classified/villa/for...    10658145   
2511   https://www.immoweb.be/en/classified/house/for...    10554963   
7853   https://www.immoweb.be/en/classified/house/for...    10649511   
11932  https://www.immoweb.be/en/classified/apartment...    10659742   
12241  https://www.immoweb.be/en/classified/house/for...    10658639   
...                                                  ...         ...   
8004   https://www.immoweb.be/en/classified/apartment...    10673932   
9623   https://www.immoweb.be/en/classified/house/for...    10668305   
15148  https://www.immoweb.be/en/classified/house/for...    10649653   
13512  https://www.immoweb.be/en/classified/villa/for...    10654779   
7238   https://www.immoweb.be/en/classified/bungalow/...    10651330   

            Type          Subtype    Price  Bedroom  Living_area  \
4930       house            villa  1300000      3.0        353.0   
2511       house            house   410000      3.0        150.0   
7853       house            house   620000      4.0        141.0   
11932      house  apartment block   285000      6.0        250.0   
12241      house            house   139000      2.0          NaN   
...          ...              ...      ...      ...          ...   
8004   apartment        apartment   349000      2.0         81.0   
9623       house            house   685000      5.0        340.0   
15148      house            house   199000      4.0        291.0   
13512      house            villa  3500000      6.0        277.0   
7238       house         bungalow   359000      3.0        137.0   

                Listing_address Postal_code              Locality  ...  \
4930              'T Helleke 12        8500              Kortrijk  ...   
2511         's Hertogendijk 95        2940              Stabroek  ...   
7853   's-Hertogenwijngaard 49,        3000                Leuven  ...   
11932                   ... ...        5500      Dinant Anseremme  ...   
12241                       / /        7350               Hensies  ...   
...                         ...         ...                   ...  ...   
8004        staakskensstraat 63        9000                  Gent  ...   
9623         stationsstraat 101        9690           Kluisbergen  ...   
15148  stationsstraat 8, gistel        8470                Gistel  ...   
13512              villapad 12,        8300                Knokke  ...   
7238            winkelstraat 24        9550  STEENHUIZE-WIJNHUIZE  ...   

      Swimming_pool  Garden Garden_area  Surface_of_land  Terrace  \
4930          False     NaN         NaN           1661.0     True   
2511          False    True         NaN            447.0     True   
7853          False    True        82.0            241.0     True   
11932         False     NaN         NaN              1.0     True   
12241         False     NaN         NaN            138.0     True   
...             ...     ...         ...              ...      ...   
8004          False     NaN         NaN              NaN     True   
9623          False     NaN         NaN           2444.0     True   
15148         False     NaN         NaN            268.0     True   
13512         False     NaN         NaN            588.0     True   
7238          False     NaN         NaN           1040.0     True   

             Kitchen Facade  Open Fire  Furnished State of the building  
4930       installed    4.0      False      False                AS_NEW  
2511       installed    3.0       True        NaN                  GOOD  
7853       installed    2.0      False      False                  GOOD  
11932  semi equipped    2.0      False      False           TO_RENOVATE  
12241      installed    2.0      False      False                  GOOD  
...              ...    ...        ...        ...                   ...  
8004             NaN    NaN      False        NaN                AS_NEW  
9623   semi equipped    NaN       True        NaN           T

After some careful lookthrough, it is apparent that duplicated entries still persist. The DataFrame object below shows the count of entries with the same address, eventhough they are assigned with unique "Listing_ID"

In [17]:
df_main_filtered["Listing_address"].value_counts()

Listing_address
None None                      2016
Maurice Devriendtlaan 21-43      99
Raymonde de Larochelaan 1        45
Heerweg-Zuid 193                 25
Rue de Lillois None              24
                               ... 
Jan Frans Stynenlei 27            1
Jan Blockxlaan 83                 1
Jan Blockxlaan 118                1
Jan Blockxlaan 114                1
winkelstraat 24                   1
Name: count, Length: 6022, dtype: int64

The removal of the duplicates listed above can be done. The script below will initiate the removal process.

###### recheck if listings are really in the same location (check on number and such)

In [18]:
df_main_filtered = df_main.sort_values(
    by=["Listing_ID"], ascending=True).drop_duplicates(
        subset=["Listing_address"], keep="last")

display(df_main_filtered.sort_values(by="Listing_address"), df_main_filtered.info())


<class 'pandas.core.frame.DataFrame'>
Index: 6022 entries, 9552 to 7952
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   URL                    6022 non-null   object 
 1   Listing_ID             6022 non-null   int64  
 2   Type                   6022 non-null   object 
 3   Subtype                6022 non-null   object 
 4   Price                  6022 non-null   object 
 5   Bedroom                5871 non-null   float64
 6   Living_area            5336 non-null   float64
 7   Listing_address        6022 non-null   object 
 8   Postal_code            6022 non-null   object 
 9   Locality               6022 non-null   object 
 10  District               6011 non-null   object 
 11  Swimming_pool          6022 non-null   bool   
 12  Garden                 2029 non-null   object 
 13  Garden_area            1084 non-null   float64
 14  Surface_of_land        3879 non-null   float64
 15  Terrac

URL  Listing_ID  \
12397  https://www.immoweb.be/en/classified/villa/for...    10658145   
2511   https://www.immoweb.be/en/classified/house/for...    10554963   
7853   https://www.immoweb.be/en/classified/house/for...    10649511   
11932  https://www.immoweb.be/en/classified/apartment...    10659742   
12231  https://www.immoweb.be/en/classified/house/for...    10658700   
...                                                  ...         ...   
8004   https://www.immoweb.be/en/classified/apartment...    10673932   
1927   https://www.immoweb.be/en/classified/house/for...    10668305   
15148  https://www.immoweb.be/en/classified/house/for...    10649653   
13512  https://www.immoweb.be/en/classified/villa/for...    10654779   
7238   https://www.immoweb.be/en/classified/bungalow/...    10651330   

            Type          Subtype    Price  Bedroom  Living_area  \
12397      house            villa  1300000      3.0        353.0   
2511       house            house   410000      3.0        150.0   
7853       house            house   620000      4.0        141.0   
11932      house  apartment block   285000      6.0        250.0   
12231      house            house   139000      2.0          NaN   
...          ...              ...      ...      ...          ...   
8004   apartment        apartment   349000      2.0         81.0   
1927       house            house   685000      5.0        340.0   
15148      house            house   199000      4.0        291.0   
13512      house            villa  3500000      6.0        277.0   
7238       house         bungalow   359000      3.0        137.0   

                Listing_address Postal_code              Locality  ...  \
12397             'T Helleke 12        8500              Kortrijk  ...   
2511         's Hertogendijk 95        2940              Stabroek  ...   
7853   's-Hertogenwijngaard 49,        3000                Leuven  ...   
11932                   ... ...        5500      Dinant Anseremme  ...   
12231                       / /        7380             Quievrain  ...   
...                         ...         ...                   ...  ...   
8004        staakskensstraat 63        9000                  Gent  ...   
1927         stationsstraat 101        9690           Kluisbergen  ...   
15148  stationsstraat 8, gistel        8470                Gistel  ...   
13512              villapad 12,        8300                Knokke  ...   
7238            winkelstraat 24        9550  STEENHUIZE-WIJNHUIZE  ...   

      Swimming_pool  Garden Garden_area  Surface_of_land  Terrace  \
12397         False     NaN         NaN           1661.0     True   
2511          False    True         NaN            447.0     True   
7853          False    True        82.0            241.0     True   
11932         False     NaN         NaN              1.0     True   
12231         False     NaN         NaN            138.0     True   
...             ...     ...         ...              ...      ...   
8004          False     NaN         NaN              NaN     True   
1927          False     NaN         NaN           2444.0     True   
15148         False     NaN         NaN            268.0     True   
13512         False     NaN         NaN            588.0     True   
7238          False     NaN         NaN           1040.0     True   

             Kitchen Facade  Open Fire  Furnished State of the building  
12397      installed    4.0      False      False                AS_NEW  
2511       installed    3.0       True        NaN                  GOOD  
7853       installed    2.0      False      False                  GOOD  
11932  semi equipped    2.0      False      False           TO_RENOVATE  
12231      installed    2.0      False      False                  GOOD  
...              ...    ...        ...        ...                   ...  
8004             NaN    NaN      False        NaN                AS_NEW  
1927   semi equipped    NaN       True        NaN           T

None

Next, it is apparent that there are entries that contains unrecognizable strings. Take a look at the DataFrame object below

In [19]:
display(df_main_filtered[df_main_filtered["Listing_address"] == "... ..."], df_main_filtered[df_main_filtered["Listing_address"] == "/ /"])

URL  Listing_ID   Type  \
11932  https://www.immoweb.be/en/classified/apartment...    10659742  house   

               Subtype   Price  Bedroom  Living_area Listing_address  \
11932  apartment block  285000      6.0        250.0         ... ...   

      Postal_code          Locality  ... Swimming_pool  Garden Garden_area  \
11932        5500  Dinant Anseremme  ...         False     NaN         NaN   

       Surface_of_land  Terrace        Kitchen Facade  Open Fire  Furnished  \
11932              1.0     True  semi equipped    2.0      False      False   

      State of the building  
11932           TO_RENOVATE  

[1 rows x 21 columns]

URL  Listing_ID   Type  \
12231  https://www.immoweb.be/en/classified/house/for...    10658700  house   

      Subtype   Price  Bedroom  Living_area Listing_address Postal_code  \
12231   house  139000      2.0          NaN             / /        7380   

        Locality  ... Swimming_pool  Garden Garden_area  Surface_of_land  \
12231  Quievrain  ...         False     NaN         NaN            138.0   

       Terrace    Kitchen Facade  Open Fire  Furnished State of the building  
12231     True  installed    2.0      False      False                  GOOD  

[1 rows x 21 columns]

Based on the observation above, it is apparent that there are entries in which the "Listing_address" contains unrecognized patterns of characters other than [a-zA-Z0-9]. It is imperative to remove them. The script below will execute the removal of such entries. 

In [20]:
df_main_filtered.drop(df_main_filtered[df_main_filtered["Listing_address"] == "... ..."].index, inplace=True)
df_main_filtered.drop(df_main_filtered[df_main_filtered["Listing_address"] == "/ /"].index, inplace=True)

display(df_main_filtered)

URL  Listing_ID  \
9552   https://www.immoweb.be/en/classified/apartment...     7893924   
9085   https://www.immoweb.be/en/classified/mixed-use...     9297070   
1498   https://www.immoweb.be/en/classified/mixed-use...     9330999   
7023   https://www.immoweb.be/en/classified/apartment...     9360253   
11437  https://www.immoweb.be/en/classified/new-real-...     9396144   
...                                                  ...         ...   
7958   https://www.immoweb.be/en/classified/apartment...    10674056   
7956   https://www.immoweb.be/en/classified/castle/fo...    10674058   
7955   https://www.immoweb.be/en/classified/house/for...    10674060   
7954   https://www.immoweb.be/en/classified/house/for...    10674061   
7952   https://www.immoweb.be/en/classified/house/for...    10674072   

                  Type             Subtype            Price  Bedroom  \
9552         apartment           apartment           325000      2.0   
9085             house  mixed use building           495000     10.0   
1498             house  mixed use building           179000      5.0   
7023         apartment           apartment           109000      2.0   
11437  apartment group     apartment group  258000 - 377500      NaN   
...                ...                 ...              ...      ...   
7958         apartment           apartment           619000      4.0   
7956             house              castle          2190000      6.0   
7955             house               house           560000      6.0   
7954             house               house            89000      1.0   
7952             house               house           325000      2.0   

       Living_area        Listing_address Postal_code            Locality  \
9552         115.0            Zeelaan 277        8670            Koksijde   
9085        1090.0      Crapaurue 108-116        4800            Verviers   
1498         216.0  Rue Crapaurue 115-117        4800            Verviers   
7023         100.0           grand rue 92        7700            Mouscron   
11437          NaN      Schipstraat 59 60        3440           Zoutleeuw   
...            ...                    ...         ...                 ...   
7958         282.0         Toekomstlaan 3        1640  SINT-GENESIUS-RODE   
7956         750.0              None None        1410            Waterloo   
7955         300.0    Rue de Belgrade 130        1060        Saint-Gilles   
7954          96.0         Rue Strivay 54        4100             Seraing   
7952         152.0      Bierbeekstraat 60        3360            Bierbeek   

       ... Swimming_pool  Garden Garden_area  Surface_of_land  Terrace  \
9552   ...         False     NaN         NaN              NaN     True   
9085   ...         False     NaN         NaN            286.0     True   
1498   ...         False    True       128.0            200.0     True   
7023   ...         False     NaN         NaN              NaN     True   
11437  ...         False     NaN         NaN              NaN     True   
...    ...           ...     ...         ...              ...      ...   
7958   ...         False    True       230.0              NaN     True   
7956   ...         False     NaN         NaN           6444.0     True   
7955   ...         False     NaN         NaN              0.0     True   
7954   ...         False     NaN         NaN             30.0     True   
7952   ...         False    True       850.0            883.0     True   

              Kitchen Facade  Open Fire  Furnished State of the building  
9552              NaN    NaN      False        NaN                AS_NEW  
9085        installed    3.0      False      False           TO_RENOVATE  
1498    not installed    2.0      False      False           TO_RENOVATE  
7023              NaN    NaN       True        NaN                   NaN  
11437             NaN    NaN      False        NaN                   NaN  
...               ...    ...        ...        ...         

The next step of the data cleaning is to make sure that the datas can be manipulated easily. To do that, there are 2 things that are needed to be done. Firstly, it is imperative to replace the "NaN" values in every entries. Secondly, it is advisable change the Dtypes accordingly so that the Dtype on each column reflects the type of data that it contains. 

The script below will address those 2 challenges.

In [21]:
df_main_filtered

URL  Listing_ID  \
9552   https://www.immoweb.be/en/classified/apartment...     7893924   
9085   https://www.immoweb.be/en/classified/mixed-use...     9297070   
1498   https://www.immoweb.be/en/classified/mixed-use...     9330999   
7023   https://www.immoweb.be/en/classified/apartment...     9360253   
11437  https://www.immoweb.be/en/classified/new-real-...     9396144   
...                                                  ...         ...   
7958   https://www.immoweb.be/en/classified/apartment...    10674056   
7956   https://www.immoweb.be/en/classified/castle/fo...    10674058   
7955   https://www.immoweb.be/en/classified/house/for...    10674060   
7954   https://www.immoweb.be/en/classified/house/for...    10674061   
7952   https://www.immoweb.be/en/classified/house/for...    10674072   

                  Type             Subtype            Price  Bedroom  \
9552         apartment           apartment           325000      2.0   
9085             house  mixed use building           495000     10.0   
1498             house  mixed use building           179000      5.0   
7023         apartment           apartment           109000      2.0   
11437  apartment group     apartment group  258000 - 377500      NaN   
...                ...                 ...              ...      ...   
7958         apartment           apartment           619000      4.0   
7956             house              castle          2190000      6.0   
7955             house               house           560000      6.0   
7954             house               house            89000      1.0   
7952             house               house           325000      2.0   

       Living_area        Listing_address Postal_code            Locality  \
9552         115.0            Zeelaan 277        8670            Koksijde   
9085        1090.0      Crapaurue 108-116        4800            Verviers   
1498         216.0  Rue Crapaurue 115-117        4800            Verviers   
7023         100.0           grand rue 92        7700            Mouscron   
11437          NaN      Schipstraat 59 60        3440           Zoutleeuw   
...            ...                    ...         ...                 ...   
7958         282.0         Toekomstlaan 3        1640  SINT-GENESIUS-RODE   
7956         750.0              None None        1410            Waterloo   
7955         300.0    Rue de Belgrade 130        1060        Saint-Gilles   
7954          96.0         Rue Strivay 54        4100             Seraing   
7952         152.0      Bierbeekstraat 60        3360            Bierbeek   

       ... Swimming_pool  Garden Garden_area  Surface_of_land  Terrace  \
9552   ...         False     NaN         NaN              NaN     True   
9085   ...         False     NaN         NaN            286.0     True   
1498   ...         False    True       128.0            200.0     True   
7023   ...         False     NaN         NaN              NaN     True   
11437  ...         False     NaN         NaN              NaN     True   
...    ...           ...     ...         ...              ...      ...   
7958   ...         False    True       230.0              NaN     True   
7956   ...         False     NaN         NaN           6444.0     True   
7955   ...         False     NaN         NaN              0.0     True   
7954   ...         False     NaN         NaN             30.0     True   
7952   ...         False    True       850.0            883.0     True   

              Kitchen Facade  Open Fire  Furnished State of the building  
9552              NaN    NaN      False        NaN                AS_NEW  
9085        installed    3.0      False      False           TO_RENOVATE  
1498    not installed    2.0      False      False           TO_RENOVATE  
7023              NaN    NaN       True        NaN                   NaN  
11437             NaN    NaN      False        NaN                   NaN  
...               ...    ...        ...        ...         

In [23]:
df_main_filtered.drop(df_main_filtered[df_main_filtered["Type"] == "apartment group"].index, inplace=True)
df_main_filtered.drop(df_main_filtered[df_main_filtered["Type"] == "house group"].index, inplace=True)

df_main_filtered["Bedroom"] = df_main_filtered["Bedroom"].fillna(0)
df_main_filtered["Garden"] = df_main_filtered["Garden"].fillna(False)
df_main_filtered["Surface_of_land"] = df_main_filtered["Surface_of_land"].fillna(0)
df_main_filtered["Kitchen"] = df_main_filtered["Kitchen"].fillna("Unknown")
df_main_filtered["Furnished"] = df_main_filtered["Furnished"].fillna("Unknown")
df_main_filtered["State of the building"] = df_main_filtered["State of the building"].fillna("UNDETERMINED")

df_main_filtered["Postal_code"] = df_main_filtered["Postal_code"].apply(lambda x: x.replace(" bs", ""))
df_main_filtered["Postal_code"] = df_main_filtered["Postal_code"].apply(lambda x: x.replace(" av", ""))
df_main_filtered["Postal_code"] = df_main_filtered["Postal_code"].apply(lambda x: x.replace(" gb", ""))
df_main_filtered["Postal_code"] = df_main_filtered["Postal_code"].apply(lambda x: x.replace(" be", ""))
dtypes = {
    "URL": "object",
    "Listing_ID": "int64", 
    "Type": "object",
    "Subtype": "category", 
    "Price": "int64",
    "Listing_address": "category",
    "Postal_code" : "int64",
    "Locality" : "category", 
    "District": "category",
    "Kitchen": "category",
    "Furnished" : "category", 
    "State of the building" : "category"}
df_main_filtered.astype(dtypes).dtypes

df_main_filtered


URL  Listing_ID  \
9552   https://www.immoweb.be/en/classified/apartment...     7893924   
9085   https://www.immoweb.be/en/classified/mixed-use...     9297070   
1498   https://www.immoweb.be/en/classified/mixed-use...     9330999   
7023   https://www.immoweb.be/en/classified/apartment...     9360253   
13860  https://www.immoweb.be/en/classified/apartment...     9812049   
...                                                  ...         ...   
7958   https://www.immoweb.be/en/classified/apartment...    10674056   
7956   https://www.immoweb.be/en/classified/castle/fo...    10674058   
7955   https://www.immoweb.be/en/classified/house/for...    10674060   
7954   https://www.immoweb.be/en/classified/house/for...    10674061   
7952   https://www.immoweb.be/en/classified/house/for...    10674072   

            Type             Subtype    Price  Bedroom  Living_area  \
9552   apartment           apartment   325000      2.0        115.0   
9085       house  mixed use building   495000     10.0       1090.0   
1498       house  mixed use building   179000      5.0        216.0   
7023   apartment           apartment   109000      2.0        100.0   
13860      house     apartment block   700000      7.0          NaN   
...          ...                 ...      ...      ...          ...   
7958   apartment           apartment   619000      4.0        282.0   
7956       house              castle  2190000      6.0        750.0   
7955       house               house   560000      6.0        300.0   
7954       house               house    89000      1.0         96.0   
7952       house               house   325000      2.0        152.0   

                     Listing_address Postal_code             Locality  ...  \
9552                     Zeelaan 277        8670             Koksijde  ...   
9085               Crapaurue 108-116        4800             Verviers  ...   
1498           Rue Crapaurue 115-117        4800             Verviers  ...   
7023                    grand rue 92        7700             Mouscron  ...   
13860  Edmond de Grimberghestraat 12        1080  Sint-Jans-Molenbeek  ...   
...                              ...         ...                  ...  ...   
7958                  Toekomstlaan 3        1640   SINT-GENESIUS-RODE  ...   
7956                       None None        1410             Waterloo  ...   
7955             Rue de Belgrade 130        1060         Saint-Gilles  ...   
7954                  Rue Strivay 54        4100              Seraing  ...   
7952               Bierbeekstraat 60        3360             Bierbeek  ...   

      Swimming_pool  Garden  Garden_area  Surface_of_land  Terrace  \
9552          False   False          NaN              0.0     True   
9085          False   False          NaN            286.0     True   
1498          False    True        128.0            200.0     True   
7023          False   False          NaN              0.0     True   
13860         False   False          NaN            407.0     True   
...             ...     ...          ...              ...      ...   
7958          False    True        230.0              0.0     True   
7956          False   False          NaN           6444.0     True   
7955          False   False          NaN              0.0     True   
7954          False   False          NaN             30.0     True   
7952          False    True        850.0            883.0     True   

              Kitchen Facade  Open Fire  Furnished State of the building  
9552          Unknown    NaN      False    Unknown                AS_NEW  
9085        installed    3.0      False      False           TO_RENOVATE  
1498    not installed    2.0      False      False           TO_RENOVATE  
7023          Unknown    NaN       True    Unknown          UNDETERMINED  
13860         Unknown    2.0      False      False                  GOOD  
...               ...    ...        ...        ...                   ...  
7958        installed    3.0    

#### General Overview from df_main_filtered

 - What kind of classified listings are there in 'Immoweb' ? 

In [29]:
classified_type = pd.DataFrame(df_main_filtered["Type"].value_counts()).reset_index()
classified_subtype = pd.DataFrame(df_main_filtered["Subtype"].value_counts()).reset_index()
print(classified_subtype)

                 Subtype  count
0                  house   2975
1              apartment   1535
2                  villa    388
3        apartment block    148
4                 duplex    126
5     mixed use building    108
6           ground floor    105
7              penthouse     83
8            flat studio     83
9                mansion     50
10  exceptional property     46
11            town house     44
12       country cottage     32
13              bungalow     31
14          service flat     29
15                  loft     20
16           manor house     14
17             farmhouse     13
18                chalet     12
19        other property     12
20               triplex      6
21                   kot      5
22                castle      4


In [30]:
labels = [x for x in classified_type["Type"]]
info= [y for y in classified_type["count"]]

types, subtypes = plt.subplots

TypeError: cannot unpack non-iterable function object

In [ ]:
price_distribution = pd.DataFrame(df_main_filtered["Price"].value_counts())

price_

Empty DataFrame
Columns: [URL, Listing_ID, Type, Subtype, Price, Bedroom, Living_area, Listing_address, Postal_code, Locality, District, Swimming_pool, Garden, Garden_area, Surface_of_land, Terrace, Kitchen, Facade, Open Fire, Furnished, State of the building]
Index: []

[0 rows x 21 columns]